Fixes #16456: Adds Puppet Path to SmartProxies

zjhuntin opened the pull request:
This is adding a new smart-proxy endpoint to allow a smart proxy to use whatever path they want for a puppet repo.

ehelms commented on app/models/katello/glue/pulp/repo.rb:
```diff
@@ -186,7 +186,7 @@ def importer_ssl_options(capsule = nil)
             :ssl_ca_cert => ::Cert::Certs.ca_cert
           }
-        elsif self.try(:redhat?) && self.content_view.default?
+        elsif redhat? && self.content_view.default?
```
This change seems from a bad rebase maybe?

ehelms commented on app/services/katello/proxy_status/pulp.rb:
```diff
@@ -5,10 +5,17 @@ module ProxyStatus
     class Pulp < ::ProxyStatus::Base
       def storage
         fetch_proxy_data do
+          require 'byebug'; byebug
```
https://www.youtube.com/watch?v=BmvLexamrmk

stbenjam commented on lib/proxy_api/pulp.rb:
```diff
@@ -11,5 +11,12 @@ def pulp_storage
+      @url += "/puppet"
+      @capsule_puppet_path ||= parse(get)
+    rescue => e
+      raise ::ProxyAPI::ProxyException.new(url, e, N_("Unable to detect pulp storage"))
```
Maybe say `s/storage/puppet status/` or something similar

stbenjam commented on app/models/katello/glue/pulp/repo.rb:
```diff
@@ -199,7 +199,7 @@ def importer_ssl_options(capsule = nil)
           }
         end
-        importer_options.merge!(:ssl_validation => verify_ssl_on_sync?) unless self.is_a?(::Katello::ContentViewPuppetEnvironment)
+        importer_options.merge(:ssl_validation => verify_ssl_on_sync?) unless self.is_a?(::Katello::ContentViewPuppetEnvironment)
```
Ditto, is this meant to be here?

zjhuntin replied:
It does not, I had added a commit by John Mitsch to make some of this work awhile ago, and had not rebased properly. About to remove that and have it working properly.

stbenjam commented:
What do you think about caching the directory in our database?

Foreman remote execution does this with SSH keys, and only reloads the key in the DB when a user refreshes smart proxy features:
https://github.com/theforeman/foreman_remote_execution/pull/55/files#diff-3f36b7c884e9cfb8f56e88ab1efe2d0bR2

stbenjam commented on app/controllers/katello/concerns/smart_proxies_controller_extensions.rb:
```diff
@@ -5,10 +5,23 @@ module SmartProxiesControllerExtensions
         alias_method_chain :action_permission, :katello
         alias_method_chain :show, :content
 
+        def capsule_puppet_path
```
This method is nearly identical to the `pulp_storage` one below, could we make it DRY'er?

stbenjam commented on app/controllers/katello/concerns/smart_proxies_controller_extensions.rb:
```diff
@@ -5,23 +5,16 @@ module SmartProxiesControllerExtensions
+          call_connection(:storage)
         end
 
         def pulp_status
```
Can this one use `call_connection` too?

beav commented:
@zjhuntin is this PR ok to test, or should I wait until WIP is removed?

zjhuntin commented:
I have updated this so it's ready to test. I think I will need to update a few of the regular tests still for it to pass, but the bulk of what real changes to be made are in here. @stbenjam @ehelms @beav @komidore64 @cfouant

ehelms commented on lib/proxy_api/pulp_node.rb:
```diff
@@ -11,5 +11,12 @@ def pulp_storage
+      @url += "/puppet"
+      @capsule_puppet_path ||= parse(get)
+    rescue => e
+      raise ::ProxyAPI::ProxyException.new(url, e, N_("Unable to detect pulp path"))
```
Should this be puppet path?

zjhuntin replied:
Yeah it should, will fix post-haste.

zjhuntin commented on config/routes/overrides.rb:
```diff
@@ -29,6 +29,7 @@ def matches?(request)
     member do
       get :pulp_storage
       get :pulp_status
+      get :capsule_puppet_path
```
I'm also not exactly sure, what this does, I more added it because it seemed like everything that was being used with functions alongside capsule_puppet_path, so I had figured it was needed. I'll try it without and see if it doesn't need to be changed.

stbenjam commented on app/lib/actions/pulp/repository/refresh.rb:
```diff
@@ -39,7 +39,8 @@ def update_or_associate_importer(capsule_id, repository, repository_details)
           concurrence do
             existing_distributors = repository_details["distributors"]
-            repository.generate_distributors(capsule_id.present?).each do |distributor|
+            capsule = capsule_id ? ::Katello::CapsuleContent.new(SmartProxy.find(capsule_id)) : nil
```
Maybe I'm just misunderstanding, but why is this a  CapsuleContent object? Seems like you're just calling capsule_content.capsule in the generate_distributor method, which is just exactly the SmartProxy object you already have here

zjhuntin replied:
It was mainly because some of the other places using the generate_distributors method already have a capsule_content object. Also, in general it also looks like many of the methods in this section of code use 'capsule' they are using a capsule_content object so I was trying to avoid confusion. Though I hear what you're saying. I'll go ahead and change it.

stbenjam commented on lib/katello/permissions/capsule_content_permissions.rb:
```diff
@@ -10,7 +10,7 @@
   permission :view_capsule_content,
             {
-              'smart_proxies' => [:pulp_storage, :pulp_status, :show_with_content]
+              'smart_proxies' => [:capsule_puppet_path, :pulp_storage, :pulp_status, :show_with_content]
```
Is this leftover? Don't see a smart_proxies#capsule_puppet_path controller method

zjhuntin replied:
It is left over, I'll get rid of it.

zjhuntin commented:
@stbenjam pushed up changes related to the comments! :bread:

stbenjam commented on app/models/katello/concerns/smart_proxy_extensions.rb:
```diff
@@ -45,6 +47,23 @@ def self.default_capsule
+      end
+
+      def update_puppet_path
+        return unless has_feature?(PULP_NODE_FEATURE)
```
I tested and it works on an external capsule, but not the Katello.  Katello's have `PULP_FEATURE` not the Node feature and would need to hit ProxyAPI::Pulp instead.

stbenjam replied:
Or, we could maybe deduplicate the two classes in ProxyAPI since they're basically identical and hit /pulp or /pulpnode depending on which feature's enabled.

zjhuntin replied:
Ahhh, okay I see now. I hadn't understood what was using the pulp feature. I'll go ahead and trying deduplicating the two.

zjhuntin commented:
@stbenjam I fixed it so it should properly update the puppet path for both Katello and capsules. I didn't feel totally comfortable in my ability to deduplicate the classes, so I opted for a different way. I also changed the content_view_puppet_environment so that it'll also use the cached puppet_path, and deleted the setting that we were using for the path in the engine file.

ehelms commented:
Looks like you got a rubocop failure here.

zjhuntin commented:
Yeah, found and fixed.

zjhuntin commented:
Tests are green!

stbenjam commented:
For my external capsule it seems to be grabbing the katello's puppet path, not the capsules.  But, the capsule sync task fails for me:

```
  worker_name: reserved_resource_worker-0@centos7-capsule-p4-nightly.example.com
  result: 
  error:
    code: PLP0034
    data:
      distributor_id: Default_Organization-Library-Test_CV
      repo_id: Default_Organization-Library-Test_CV
      summary: 'failed to create destination directory: [Errno 13] Permission denied:
        ''/etc/puppetlabs/code/environments/KT_Default_Organization_Library_Test_CV_2'''
    description: The distributor Default_Organization-Library-Test_CV indicated a
      failed response when publishing repository Default_Organization-Library-Test_CV.
```

But that is not the path my smart proxy uses:

```
irb(main):002:0> SmartProxy.friendly.find('centos7-capsule-p4-nightly.example.com')
=> #<SmartProxy id: 2, name: "centos7-capsule-p4-nightly.example.com", url: "https://centos7-capsule-p4-nightly.example.com:909...", created_at: "2016-09-07 18:06:18", updated_at: "2016-09-07 18:48:51", expired_logs: "0", puppet_path: "/etc/puppet/environments">
```

zjhuntin commented:
@stbenjam Sorry about that. I had made a bad assumption at a certain point that I could use the internal capsule's puppet path which was leading to the error that you had seen. I think that I have that solved now. If you are going to retry this use a new content view as I think it may have saved the path that it was trying to publish the content to in the database. Also, I put in a fail clause when trying to publish the puppet content locally because doing katello:reset in a development environment wipes the internal capsule, so it wouldn't be possible to do so.

stbenjam commented:
>  If you are going to retry this use a new content view as I think it may have saved the path that it was trying to publish the content to in the database. 

Ew :-(

I guess that will affect upgrades, right? If a user upgrades the local katello or a capsule to Puppet 4, or otherwise somehow changes the path, the paths stay hardcoded?

beav commented:
Can you do something like this to refresh?

```ruby
        Katello::Repository.find_each do |repo|
          ForemanTasks.sync_task(::Actions::Katello::Repository::RefreshRepository, repo)
        end
```

stbenjam commented:
With the latest changes, it tries to publish to the correct paths on my Katello and the external capsule :+1: :tada: 

I think @beav's idea is good, maybe we can do that whenever the puppet_path changes.

beav commented:
I don't think it is too expensive to do the refresh. I don't know if its better to do it sync or async, I guess async is better from a user experience perspective.

stbenjam commented:
So, @zjhuntin and I tested after some help from Justin, and turns out we don't need to do it.  Changing the path on the capsule, and publishing an existing CV makes the content go to the new path just fine. :tada:

ACK from me. I'm not super familiar with the content parts of Katello, so it'd be great if someone else would review and/or test too.

beav commented:
The changes look good to me as well, I see in `repos_needing_distributor_updates` where it catches the path update.

What was the issue you hit originally when publishing an existing CV?

zjhuntin commented:
@beav I thought it was trying to publish it to the wrong place again. But I think I was mistaken, @jlsherrill  showed me in my dynflow tasks that it did see a difference with what it thought the configuration was, and what I was telling it, then changed appropriately.

beav commented:
@zjhuntin ah, cool. ACK from me as well from a code review perspective, the changes are straightforward.

johnpmitsch commented:
ACK

johnpmitsch commented:
Thanks @zjhuntin!
